src/core/column/cut.h

```diff
@@ -0,0 +1,131 @@
+//------------------------------------------------------------------------------
+// Copyright 2019-2020 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#ifndef dt_COLUMN_CUT_h
+#define dt_COLUMN_CUT_h
+#include <memory>
+#include "column/func_unary.h"
+#include "column/virtual.h"
+#include "stype.h"
+
+namespace dt {
+
+
+/**
+  * Virtual column for data binning.
+  */
+template <typename T_elems, typename T_stats>
+class Cut_ColumnImpl : public Virtual_ColumnImpl {
+  private:
+    Column col_dbl_;
+    double bin_factor_, bin_shift_;
+
+  public:
+    Cut_ColumnImpl(Column col, size_t bins)
+      : Virtual_ColumnImpl(col.nrows(), col.stype()),
+      bin_factor_(dt::NA_F8),
+      bin_shift_(dt::NA_F8)
+    {
+      col_dbl_ = Column(new dt::FuncUnary2_ColumnImpl<T_elems, double>(
+                   Column(col),
+                   [](T_elems x, bool x_isvalid, double* out) {
+                     *out = static_cast<double>(x);
+                     return x_isvalid && _isfinite(x);
+                   },
+                   col.nrows(),
+                   dt::SType::FLOAT64
+                 ));
```

Comment:
Looks like you're trying to convert the column into float64 type here -- it's easier (and potentially more efficient) to use `col.cast_inplace(SType::FlOAT64)`.

I know you're converting infinities to NAs here too, but that cost should not be incurred unless actually needed, because columns with infinities are extremely rare. So, the simplest approach would be to ignore them completely; while a more advanced one is to proceed as if there are no infinities, but if min/max operations below return infinities, then treat this as a special case.